lib/Middleware/PermissionMiddleware.php

```diff
@@ -98,21 +101,78 @@ protected function checkPermission(RequirePermission $attribute): void {
-					throw new PermissionError(sprintf('User %s cannot manage context %d',
-						$this->userId, $nodeId
-					));
-				}
+		match ($attribute->getPermission()) {
+			Application::PERMISSION_MANAGE => $this->assertManagePermission($isContext, $nodeType, $nodeId),
+			Application::PERMISSION_CREATE => $this->assertCreatePermissions($nodeType, $nodeId),
+			Application::PERMISSION_UPDATE => $this->assertUpdatePermissions($nodeType, $nodeId),
+			Application::PERMISSION_DELETE => $this->assertDeletePermissions($nodeType, $nodeId),
+			Application::PERMISSION_ALL => $this->assertManagePermission($isContext, $nodeType, $nodeId)
+				&& $this->assertCreatePermissions($nodeType, $nodeId)
+				&& $this->assertUpdatePermissions($nodeType, $nodeId)
+				&& $this->assertDeletePermissions($nodeType, $nodeId),
+		};
+	}
+
+	/**
+	 * @throws PermissionError
+	 */
+	private function assertCreatePermissions(int $nodeType, int $nodeId): bool {
+		if (!$this->permissionsService->canCreateRowsById($nodeType, $nodeId, $this->userId)) {
+			throw new PermissionError(sprintf('User %s cannot create rows on node %d (type %d)',
+				$this->userId, $nodeId, $nodeType
+			));
+		}
+		return true;
+	}
+
+	/**
+	 * @throws PermissionError
+	 */
+	private function assertUpdatePermissions(int $nodeType, int $nodeId): bool {
+		if (
+			($nodeType === Application::NODE_TYPE_TABLE && !$this->permissionsService->canUpdateRowsByTableId($nodeId, $this->userId))
+			|| ($nodeType === Application::NODE_TYPE_VIEW && !$this->permissionsService->canUpdateRowsByViewId($nodeId, $this->userId))
+		) {
+			throw new PermissionError(sprintf('User %s cannot update rows on node %d (type %d)',
+				$this->userId, $nodeId, $nodeType
+			));
+		}
+		return true;
+	}
+
+	/**
+	 * @throws PermissionError
+	 */
+	private function assertDeletePermissions(int $nodeType, int $nodeId): bool {
+		if (
+			($nodeType === Application::NODE_TYPE_TABLE && !$this->permissionsService->canDeleteRowsByTableId($nodeId, $this->userId))
+			|| ($nodeType === Application::NODE_TYPE_VIEW && !$this->permissionsService->canDeleteRowsByViewId($nodeId, $this->userId))
+		) {
+			throw new PermissionError(sprintf('User %s cannot delete rows on node %d (type %d)',
+				$this->userId, $nodeId, $nodeType
+			));
+		}
+		return true;
+	}
+
+	/**
+	 * @throws PermissionError
+	 */
+	private function assertManagePermission(bool $isContext, int $nodeType, int $nodeId): bool {
+		if (!$isContext) {
+			if (!$this->permissionsService->canManageNodeById($nodeType, $nodeId, $this->userId)) {
+				throw new PermissionError(sprintf('User %s cannot manage node %d (type %d)',
+					$this->userId, $nodeId, $nodeType
+				));
+			}
+		} else {
+			if (!$this->permissionsService->canManageContextById($nodeId, $this->userId)) {
+				throw new PermissionError(sprintf('User %s cannot manage context %d',
+					$this->userId, $nodeId
+				));
 			}
 		}
+		return true;
```

Comment:
Is there a use case for returning true in the assert-Methods if we throw anways in case of no permission?

Reply:
Otherwise I could not chain them in the match block: https://github.com/nextcloud/tables/pull/1161/files#diff-58a7c8ba568eef76405b8471729c4b735def0967160958d5ab0105ddb7afb34eR109-R112

could turn it into switch construction instead of course. I find the match more compact and readable.

Reply:
Ah ok, that makes sense then 👍